File: leetcode/453_minimum-moves-to-equal-array-elements/453.py
class Solution(object):
    # Execeed Time Limit
    def minMoves(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        m, n, nums = 0, len(nums), sorted(nums)
        while nums[0] != nums[-1]:
            i = n - 1
            while nums[i] == nums[-1]: i -= 1
            k = nums[i+1] - nums[i]
            for j in range(n):
                if j != i+1: nums[j] += k
            m += k
        return m, nums


if __name__ == '__main__':
    s = Solution()
    print (s.minMoves([1,2,3,4]))
    print (s.minMoves([3,3,3,4]))
    print (s.minMoves([3,3,6,6]))

File: leetcode/453_minimum-moves-to-equal-array-elements/453.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

using namespace std;

class Solution {
public:
    int minMoves(vector<int>& nums) {
        int moves = 0;
        vector<int>::const_iterator minIt = min_element(nums.begin(), nums.end());
        for (vector<int>::const_iterator it = nums.begin();
             it != nums.end(); ++it) {
            moves += *it - *minIt;
        }
        return moves;
    }
};

int main(void) {

    Solution s;
    vector<int> nums{1,2,3,4};
    cout << s.minMoves(nums) << endl;

    return 0;
}
